File: JavaScript/isAnagram.js
// 242. Valid Anagram

// Given two strings s and t, return true if t is an anagram of s, and false otherwise.

// An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

// Example 1:

// Input: s = "anagram", t = "nagaram"
// Output: true

// Example 2:

// Input: s = "rat", t = "car"
// Output: false

// Constraints:

//     1 <= s.length, t.length <= 5 * 104
//     s and t consist of lowercase English letters.
// Follow up: What if the inputs contain Unicode characters? How would you adapt your solution to such a case?

var isAnagram = function(s, t) {
    if (s.length !== t.length) return false;

    const store = new Map();

    for (let i = 0; i < s.length; i++) {
        const charS = s.charAt(i);
        const charT = t.charAt(i);

        store.set(charS, (store.get(charS) || 0) + 1);
        store.set(charT, (store.get(charT) || 0) - 1);
    }

    for (let count of store.values()) {
        if (count !== 0) return false;
    }

    return true;
};
console.log(isAnagram("anagram","nagaram"));

File: Accenture_Coding_Questions/IsAnagram.py
# 3. Write a function to validate if the provided two strings are anagrams or not. If the two strings
# are anagrams, then return ‘yes’. Otherwise, return ‘no’.
# Input:
# Input 1: 1st string
# Input 2: 2nd string
# Output:
# (If they are anagrams, the function will return ‘yes’. Otherwise, it will return ‘no’.)
# Example
# Input 1: Listen
# Input 2: Silent
# Output:
# Yes
# Explanation
# Listen and Silent are anagrams (an anagram is a word formed by rearranging the letters of the
# other word).

def IsAnagram(str1,str2):
    str1 = str1.replace(" ", "").lower()
    str2 = str2.replace(" ", "").lower()
    
    if(len(str1)!=len(str2)):
        return "No"
    
    char_count = {}
    for char in str1:
        char_count[char] = char_count.get(char, 0) + 1
    for char in str2:
        if char in char_count and char_count[char] > 0:
            char_count[char] -= 1
        else:
            return "No"
    return "Yes"

print(IsAnagram("Listen", "Silent"))